api/handlers_test.go

```diff
@@ -0,0 +1,139 @@
+			})
+		})
+	})
+
+	Convey("Given a valid create session request to /session with body with all elements", t, func() {
+		mockSessions := &SessionsMock{
+			NewFunc: func(email string) (*session.Session, error) {
+				return &session.Session{
+					ID:    "1234",
+					Email: email,
+				}, nil
+			},
+		}
+		sessionHandler := CreateSessionHandlerFunc(mockSessions)
+
+		sess := session.NewSessionDetails{
+			Email: "test@test.com",
+		}
+		sessJSON, err := json.Marshal(sess)
+		So(err, ShouldBeNil)
+
+		req := httptest.NewRequest("POST", "/session", strings.NewReader(string(sessJSON)))
+		resp := httptest.NewRecorder()
+
+		Convey("When the request is handled by the router", func() {
+			sessionHandler.ServeHTTP(resp, req)
+
+			Convey("Then the expected success response is returned", func() {
+				So(resp.Code, ShouldEqual, 201)
+				So(resp.Header().Get("Content-Location"), ShouldEqual, fmt.Sprintf("/session/1234"))
+				So(mockSessions.NewCalls(), ShouldHaveLength, 1)
+				So(mockSessions.NewCalls()[0].Email, ShouldEqual, sess.Email)
+			})
+		})
+	})
+
+	Convey("Given a bad request to /session", t, func() {
+		mockSessions := &SessionsMock{}
+		sessionHandler := CreateSessionHandlerFunc(mockSessions)
+
+		req := httptest.NewRequest("POST", "/session", strings.NewReader("this is not json"))
+		resp := httptest.NewRecorder()
+
+		Convey("When the request is handled by the router", func() {
+			sessionHandler.ServeHTTP(resp, req)
+
+			Convey("Then return an error response", func() {
+				So(resp.Code, ShouldEqual, http.StatusInternalServerError)
+				So(mockSessions.NewCalls(), ShouldHaveLength, 0)
+			})
+		})
+	})
+
+	Convey("Given a request to /session with a body with missing elements", t, func() {
+		mockSessions := &SessionsMock{
+			NewFunc: func(email string) (*session.Session, error) {
+				return &session.Session{
+					ID:    "1234",
+					Email: "",
+				}, nil
+			},
+		}
+		sessionHandler := CreateSessionHandlerFunc(mockSessions)
+
+		sess := session.NewSessionDetails{
+			Email: "",
+		}
+		sessJSON, err := json.Marshal(sess)
+		So(err, ShouldBeNil)
+
+		req := httptest.NewRequest("POST", "/session", strings.NewReader(string(sessJSON)))
+		resp := httptest.NewRecorder()
+
+		Convey("When the request is handled by the router", func() {
+			sessionHandler.ServeHTTP(resp, req)
+
+			Convey("Then return an error response", func() {
+				So(resp.Code, ShouldEqual, http.StatusBadRequest)
+				So(mockSessions.NewCalls(), ShouldHaveLength, 0)
+			})
+		})
+	})
+
+	Convey("Given a valid create session request but UUID fails to generate", t, func() {
+		mockSessions := &SessionsMock{
+			NewFunc: func(email string) (*session.Session, error) {
+				return nil, errors.New("unable to generate id")
+			},
+		}
+		sessionHandler := CreateSessionHandlerFunc(mockSessions)
+
+		sess := session.NewSessionDetails{
+			Email: "test@test.com",
+		}
+		sessJSON, err := json.Marshal(sess)
+		So(err, ShouldBeNil)
```

Comment:
We've repeated this code 3+ times now which I think means you could probably pull this out into a helper func. A general rule of thumb I follow is - if you duplicate it 3 times consider abstracting it.